webrender_api/src/display_item.rs

```diff
@@ -865,3 +865,123 @@ impl ExternalScrollId {
+/// the amount to adjust the gradient line start and stop
+fn normalize_stops(stops: &mut Vec<GradientStop>, extend_mode: ExtendMode) -> (f32, f32) {
+    assert!(stops.len() >= 2);
+
+    let first = *stops.first().unwrap();
+    let last = *stops.last().unwrap();
+
+    assert!(first.offset <= last.offset);
+
+    let stops_delta = last.offset - first.offset;
+
+    if stops_delta > 0.000001 {
+        for stop in stops {
+            stop.offset = (stop.offset - first.offset) / stops_delta;
+        }
+
+        (first.offset, last.offset)
+    } else {
+        // We have a degenerate gradient and can't accurately transform the stops
+        // what happens here depends on the repeat behavior, but in any case
+        // we reconstruct the gradient stops to something simpler and equivalent
+        stops.clear();
+
+        match extend_mode {
+            ExtendMode::Clamp => {
+                // This gradient is two colors split at the offset of the stops,
+                // so create a gradient with two colors split at 0.5 and adjust
+                // the gradient line so 0.5 is at the offset of the stops
+                stops.push(GradientStop { color: first.color, offset: 0.0, });
+                stops.push(GradientStop { color: first.color, offset: 0.5, });
+                stops.push(GradientStop { color: last.color, offset: 0.5, });
+                stops.push(GradientStop { color: last.color, offset: 1.0, });
+
+                let offset = last.offset;
+
+                (offset - 0.5, offset + 0.5)
+            }
+            ExtendMode::Repeat => {
+                // A repeating gradient with stops that are all in the same
+                // position should just display the last color. I believe the
+                // spec says that it should be the average color of the gradient,
+                // but this matches what Gecko and Blink does
+                stops.push(GradientStop { color: last.color, offset: 0.0, });
+                stops.push(GradientStop { color: last.color, offset: 1.0, });
+
+                (0.0, 1.0)
+            }
+        }
+    }
+}
+
+impl Gradient {
+    /// Create a normalized linear gradient.
+    ///
+    /// The stops are normalized in-place.
+    pub fn new(
+        start_point: LayoutPoint,
+        end_point: LayoutPoint,
+        stops: &mut Vec<GradientStop>,
+        extend_mode: ExtendMode,
+    ) -> Gradient {
+        let (start_offset, end_offset) = normalize_stops(stops, extend_mode);
+        let start_to_end = end_point - start_point;
+
+        Gradient {
+            start_point: start_point + start_to_end * start_offset,
+            end_point: start_point + start_to_end * end_offset,
+            extend_mode,
+        }
+    }
+}
+
+impl RadialGradient {
+    /// Create a normalized radial gradient.
+    ///
+    /// The stops are normalized in-place.
+    pub fn new(
+        center: LayoutPoint,
+        radius: LayoutSize,
+        stops: &mut Vec<GradientStop>,
```

Comment:
It feels to me that the concept you really want to expose is `GradientBuilder`. This object would be able to manage `Vec<GradientStop>` internally and produce `Gradient/RadialGradient` objects.

Reply:
I am not entirely sure how the `GradientBuilder` API should look like. Can you write down the type and function signatures if you have time?

Reply:
sure!
```rust
pub struct GradientBuilder {
  stops: Vec<GradientStop>,
  // something else?
}
impl GradientBuidler {
  fn normalize(&mut self) {...}
  pub fn gradient(&mut self, points: Range<LayoutPoint>, extend_mode: ExtendMode) -> Gradient {...}
  pub fn radial_gradient(&mut self, center: LayoutPoint, radius: LayoutSize, extend_mode: ExtendMode) -> RadialGradient {..}
}
```

Reply:
How do I put stops "into" the builder and how can I get the vector "out" again?

Reply:
well, by writing code/methods doing so :)
If this is unclear of confusing to you, we can proceed as is. I'm just thinking that this code is asking for a builder like that, given how it considers the ownership of `Vec<GradientStop>` at each step.

Reply:
> well, by writing code/methods doing so :)

Sure. It's just that I am not very good at (Rust) API design. I will just add those methods as I think they work best.